Allow Area2D and 3D mouse events without collision layer

KoBeWi opened the pull request:
Resolves #11159
And consequently reverts #28608

`_intersect_point_impl()` isn't used for anything else than mouse picking, so making it ignore layers is safe.

KoBeWi committed:
Allow Area2D and 3D mouse events without a collision layer

Co-authored-by: madmiraal <madmiraal@users.noreply.github.com>

akien-mga commented:
How is the behavior in Area3D btw?

KoBeWi commented:
3D ignores layers too.

madmiraal commented on servers/physics_2d/space_2d_sw.cpp:
```diff
@@ -34,8 +34,8 @@
 #include "core/os/os.h"
 #include "core/pair.h"
 #include "physics_server_2d_sw.h"
-_FORCE_INLINE_ static bool _can_collide_with(CollisionObject2DSW *p_object, uint32_t p_collision_mask, bool p_collide_with_bodies, bool p_collide_with_areas) {
-	if (!(p_object->get_collision_layer() & p_collision_mask)) {
```
If we're updating this, then we should add the missing blank line at the end of the includes.

madmiraal reviewed (changes requested):
It fixes #11159 in 2D, but, to echo @akien-mga's question, it is still broken in 3D; both Godot and Bullet physics. This PR should be extended to fix 3D Godot and Bullet physics too to make it all consistent. Otherwise, I think it's a good solution. :+1:

KoBeWi commented:
> it is still broken in 3D; both Godot and Bullet physics

It's not, I tested it.

madmiraal commented:
I can't test it on 4.0, because master doesn't currently work on my computer, but I created a [3.2 backport](https://github.com/madmiraal/godot/tree/PR40193-3.2) (although it's not necessary since the PR doesn't change any 3D code), and tested using this simple project:
[11159.zip](https://github.com/godotengine/godot/files/4890238/11159.zip)

Moving the mouse over and off the cube should trigger signals that print "Mouse entered" and "Mouse exited" messages. Without a `Collision Layer` selected, nothing is printed; with a `Collision Layer` selected, the messages are printed. This happens with both Godot and Bullet physics.

KoBeWi commented:
Ok, my bad. For whatever reason the layer in my scene had a value of 2147483648 and it was showing in the inspector as empty :I

I looked into the 3D code and it looks more complicated, but I'll try to fix it too.

akien-mga commented:
> For whatever reason the layer in my scene had a value of 2147483648 and it was showing in the inspector as empty :I

Sounds like uninitialized memory.

KoBeWi commented:
Ok, I managed to fix it for GodotPhysics, but I have no idea how to do it for Bullet. It could be done in another PR, by someone who is more knowledgeable on that regard. (Also the issue mentioned specifically 2D, so technically it is resolved >_>)

madmiraal commented:
> I have no idea how to do it for Bullet. It could be done in another PR, by someone who is more knowledgeable on that regard.

@KoBeWi I've sent you a [PR](https://github.com/KoBeWi/godot/pull/1) for the same solution in Bullet physics.

KoBeWi commented:
@madmiraal Thanks for the code! Added it and rebased. I thought you'd be credited automatically, but seems like git removed you from the commit. I could manually add you as co-author if you want.

madmiraal commented:
> I could manually add you as co-author if you want.

Not essential, but while you're updating the commit message to "Allow Area2D **and 3D** mouse events without a collision layer" that would be nice. :-)

akien-mga commented:
Thanks!

akien-mga commented:
It changes behavior but I'd say it's a good change for 3.2.x too, do you agree?

madmiraal commented:
I agree. Although it is possible currently to disable an Area's pickability by removing it from all layers, the more logical approach is to disable pickable. It may be a long standing bug, but I don't think fixing it can be regarded as breaking compat.

@KoBeWi I have a ready-made [backport](https://github.com/madmiraal/godot/tree/PR40193-3.2) (with you as co-author) which I used for testing, ready for creating a PR; unless you want to do it.

KoBeWi commented:
> ready for creating a PR; unless you want to do it.

Go ahead.

akien-mga commented:
Reverting as per https://github.com/godotengine/godot/issues/40705#issuecomment-664554271, to fix regression #40705.
